### PR TITLE
Feature/myst-470 pass vpn options structure

### PR DESCRIPTION
**Pull request by tadovas**

### CODE REVIEW

File: cmd/commands/server/command_server.go [hunk 1]
```diff
@@ -30,10 +31,9 @@ type Command struct {
 	dialogWaiterFactory func(identity identity.Identity) communication.DialogWaiter
 	dialogWaiter        communication.DialogWaiter
 
-	sessionManagerFactory func(serverIP string) session.Manager
+	sessionManagerFactory func(primitives *tls.Primitives, serverIP string) session.Manager
 
-	vpnServerFactory func(sessionManager session.Manager, serviceLocation dto_discovery.Location,
-		providerID identity.Identity, callback state.Callback) *openvpn.Server
+	vpnServerFactory func(sessionManager session.Manager, primitives *tls.Primitives, callback state.Callback) *openvpn.Server
```
Comment (zolia) [inline cmd/commands/server/command_server.go]: I would change naming for 'callback' -> serverStateCB or stateCallBack

File: openvpn/client.go [hunk 2]
```diff
@@ -28,20 +31,23 @@ func NewClient(openvpnBinary string, config *ClientConfig, directoryRuntime stri
...
-	return func() *ClientConfig {
-		// (Re)generate required security primitives before openvpn start
-		vpnClientConfig := NewClientConfig(
+func (generator ClientConfigGenerator) ProvideServiceConfig() (session.VPNConfig, error) {
+	return generator(), nil
+}
+
+// NewClientConfigGenerator returns function generating config params for remote client
+func NewClientConfigGenerator(primitives *tls.Primitives, vpnServerIP string, port int, protocol string) ClientConfigGenerator {
+	return func() session.VPNConfig {
```
Comment (zolia) [inline openvpn/client.go]: Shouldn't this better be *reference?

File: openvpn/config.go [hunk 3]
```diff
@@ -1,17 +1,20 @@
 package openvpn
 
 import (
+	"path/filepath"
 	"strconv"
 )
 
-func NewConfig() *Config {
+func NewConfig(configdir string) *Config {
```
Comment (zolia) [inline openvpn/config.go]: configdir -> configDir

Comment (zolia) [reply openvpn/config.go]: or configDirectory

File: openvpn/config.go [hunk 4]
```diff
@@ -48,16 +51,16 @@ func (c *Config) SetDevice(deviceName string) {
 }
 
 func (c *Config) SetTLSCACertificate(caFile string) {
```
Comment (zolia) [inline openvpn/config.go]: undocumented public method

File: openvpn/config_validator.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+	case
+		"udp",
+		"tcp":
+		return nil
+	}
+	return errors.New("invalid protocol: " + config.RemoteProtocol)
+}
+
+func validPort(config session.VPNConfig) error {
+	if config.RemotePort > 65535 || config.RemotePort < 1 {
```
Comment (zolia) [inline openvpn/config_validator.go]: should be 'config.RemotePort < 1024', first 1023 ports are reserved for superUser processes and cannot be used with userSpace processes.

File: openvpn/config_validator.go [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+		return errors.New("invalid port range")
```
Comment (zolia) [inline openvpn/config_validator.go]: I would print more verbose message: "invalid port range, should fall within 1024 .. 65535 range"

File: openvpn/config_validator.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	}
+	return nil
+}
+
+func validIPFormat(config session.VPNConfig) error {
+	parsed := net.ParseIP(config.RemoteIP)
+	if parsed == nil {
+		return errors.New("unable to parse ip address" + config.RemoteIP)
```
Comment (zolia) [inline openvpn/config_validator.go]: ""unable to parse IP address: " - upcase and space after message.

File: openvpn/config_validator.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	if len(key) != 512 {
+		return errors.New("invalid key length")
+	}
+	return nil
+}
+
+func validCACertificate(config session.VPNConfig) error {
+	pemBlock, _ := pem.Decode([]byte(config.CACertificate))
+	if pemBlock.Type != "CERTIFICATE" {
+		return errors.New("invalid ca. Certificate block expected")
```
Comment (zolia) [inline openvpn/config_validator.go]: ca -> CA certificate

File: openvpn/server.go [hunk 9]
```diff
@@ -21,26 +19,23 @@ func NewServer(openvpnBinary string, generateConfig ServerConfigGenerator, direc
...
+			port,
 			protocol,
 		)
-		return vpnServerConfig, err
+		return vpnServerConfig
 	}
 }
 
-// Server structure describes openvpn server
+// ServerCertificate structure describes openvpn server
```
Comment (zolia) [inline openvpn/server.go]: should be 'Server'

File: session/dto.go [hunk 10]
```diff
@@ -6,17 +6,29 @@ import (
...
 
+//SessionDto structure represents session information data within session creation response (session id and configuration options for underlaying service type)
 type SessionDto struct {
 	ID     SessionID `json:"id"`
-	Config string    `json:"config"`
+	Config VPNConfig `json:"config"`
+}
+
+//VPNConfig structure represents VPN configuration options for given session
+type VPNConfig struct {
```
Comment (Waldz) [inline session/dto.go]: Are we mixing Openvpn specific config session into `session.config`?

File: openvpn/tls/tls_test.go [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package tls
+
+import (
+	"github.com/mysterium/node/identity"
+	"github.com/mysterium/node/service_discovery/dto"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+const fakeRunDir = "testdataoutput"
```
Comment (zolia) [inline openvpn/tls/tls_test.go]: not used anymore

File: session/dto.go [hunk 12]
```diff
@@ -6,17 +6,29 @@ import (
 
 const endpointSessionCreate = communication.RequestEndpoint("session-create")
 
+//SessionCreateRequest structure represents message from service consumer to initiate session for given proposal id
```
Comment (zolia) [inline session/dto.go]: space?

File: cmd/commands/server/factory.go [hunk 13]
```diff
@@ -79,16 +79,16 @@ func NewCommandWith(
...
 
 			return openvpn_session.NewManager(
-				clientConfigGenerator,
+				openvpn_session.ServiceConfigProvider(clientConfigGenerator),
 				&session.UUIDGenerator{},
 			)
 		},
-		vpnServerFactory: func(manager session.Manager, serviceLocation dto.Location, providerID identity.Identity, callback state.Callback) *openvpn.Server {
-			serverConfigGenerator := openvpn.NewServerConfigGenerator(options.DirectoryRuntime, serviceLocation, providerID, options.Protocol)
+		vpnServerFactory: func(manager session.Manager, primitives *tls.Primitives, callback state.Callback) *openvpn.Server {
```
Comment (Waldz) [inline cmd/commands/server/factory.go]: Too long lines :/

File: openvpn/factory.go [hunk 14]
```diff
@@ -34,16 +41,9 @@ func NewServerConfig(
...
-) *ClientConfig {
-	config := ClientConfig{NewConfig()}
-	config.SetClientMode(remote, 1194)
-	config.SetProtocol(protocol)
-	config.SetTLSCACertificate(caCertPath)
-	config.SetTLSCrypt(tlsCryptKeyPath)
+func newClientConfig(configDir string) *ClientConfig {
+	config := ClientConfig{NewConfig(configDir)}
+
 	config.RestrictReconnects()
```
Comment (Waldz) [inline openvpn/factory.go]: Strange that 2 values are here, and others in `NewClientConfigFromSession `

File: openvpn/session/manager.go [hunk 15]
```diff
@@ -2,34 +2,39 @@ package session
...
 // NewManager returns session manager which maintains a map of session id -> session
-func NewManager(clientConfigGenerator openvpn.ClientConfigGenerator, idGenerator session.Generator) *manager {
+func NewManager(serviceConfigProvider ServiceConfigProvider, idGenerator session.Generator) *manager {
 	return &manager{
-		idGenerator:          idGenerator,
-		generateClientConfig: clientConfigGenerator,
-		sessionMap:           make(map[session.SessionID]session.Session),
-		creationLock:         sync.Mutex{},
+		idGenerator:    idGenerator,
+		configProvider: serviceConfigProvider,
```
Comment (Waldz) [inline openvpn/session/manager.go]: SessionManager became not openvpn-specific, lets move out of package `openvpn/session`

**Comment by tadovas** [comment]
Fixes ready. Ping @zolia @Waldz

File: openvpn/factory.go (hunk 14 above)
Comment (tadovas) [reply openvpn/factory.go]: This function is bigger - the diff only shows two lines

File: cmd/commands/server/factory.go (hunk 13 above)
Comment (tadovas) [reply cmd/commands/server/factory.go]: Reformated :) however a notice added to TODO box - cleanup server factory

File: openvpn/server_config.go [hunk 16]
```diff
@@ -12,7 +12,7 @@ func (c *ServerConfig) SetServerMode(port int, network, netmask string) {
 
 func (c *ServerConfig) SetTLSServer() {
 	c.setFlag("tls-server")
-	c.AddOptions(OptionFile("dh", "none"))
+	c.AddOptions(OptionFile("dh", "none", "none"))
```
Comment (Waldz) [inline openvpn/server_config.go]: Are you writing "none" to file content? no good

File: cmd/commands/server/factory.go [hunk 17]
```diff
@@ -79,16 +79,26 @@ func NewCommandWith(
 			)
 		},
 
-		sessionManagerFactory: func(vpnServerIP string) session.Manager {
-			clientConfigGenerator := openvpn.NewClientConfigGenerator(options.DirectoryRuntime, vpnServerIP, options.Protocol)
+		sessionManagerFactory: func(primitives *tls.Primitives, vpnServerIP string) session.Manager {
```
Comment (Waldz) [inline cmd/commands/server/factory.go]: We are creating several instances of `sessionManager`, but it does not state anymore. 
Add TODO for this problem

**Comment by tadovas** [comment]
@Waldz comments addresed.

File: cmd/commands/server/factory.go (hunk 17 above)
Comment (tadovas) [reply cmd/commands/server/factory.go]: Commented

File: openvpn/server_config.go (hunk 16 above)
Comment (tadovas) [reply openvpn/server_config.go]: Done

**Pull request merged**